pulsar-metadata/src/main/java/org/apache/pulsar/metadata/bookkeeper/PulsarRegistrationClient.java

```diff
@@ -153,4 +212,56 @@ private static Set<BookieId> convertToBookieAddresses(List<String> children) {
         }
         return newBookieAddrs;
     }
+
+    @Override
+    public CompletableFuture<Versioned<BookieServiceInfo>> getBookieServiceInfo(BookieId bookieId) {
+        // this method cannot perform blocking calls to the MetadataStore
+        // or return a CompletableFuture that is completed on the MetadataStore main thread
+        // this is because there are a few cases in which some operations on the main thread
+        // wait for the result. This is due to the fact that resolving the address of a bookie
+        // is needed in many code paths.
+        Versioned<BookieServiceInfo> resultFromCache = bookieServiceInfoCache.get(bookieId);
+        if (log.isDebugEnabled()) {
+            log.debug("getBookieServiceInfo {} -> {}", bookieId, resultFromCache);
+        }
+        if (resultFromCache != null) {
+            return CompletableFuture.completedFuture(resultFromCache);
+        } else {
+            return FutureUtils.exception(new BKException.BKBookieHandleNotAvailableException());
+        }
+    }
+
+    public CompletableFuture<Versioned<BookieServiceInfo>> readBookieServiceInfoAsync(BookieId bookieId) {
+        String asWritable = bookieRegistrationPath + "/" + bookieId;
+        return bookieServiceInfoMetadataCache.get(asWritable)
```

Comment:
You refer to this `get` could be a blocking call.
My best guess is when the key is loading for the first time. 
But the key call path should be like :
- org.apache.pulsar.metadata.cache.impl.MetadataCacheImpl#get
- com.github.benmanes.caffeine.cache.AsyncLoadingCache#get
- org.apache.pulsar.metadata.cache.impl.MetadataCacheImpl#readValueFromStore

and all of these should not contains any blocking operations.

Reply:
`AsyncLoadingCache#get`

this is nasty, I know, because even if it seems "async" the "completion" of the CompletableFuture is executed on the main metadata store thread.

In another operation that runs on the same thread you cannot wait for the completion of a CompletableFuture that wills be only completed on the same thread.

Reply:
I see, the THREAD (main-EventThread) is not blocked at this `get` method actually. The `get` method returns a CF without blocking, but the completion of the CF depends on the THREAD, and the THREAD is blocked at the completions of the CF. Thus the deadlock.